Transactions: Feature complete

asqui opened the pull request:
This PR picks up where #1 left off.

I think this is now feature-complete with respect to the design that we discussed.

Let me know your thoughts.

asqui committed:
HACK: Adapt find_library() call to find libpq on Windows

This will need to be made conditional in some way.

asqui committed:
(not fully working) Include a portable version of endian.h

Source: https://github.com/mikepb/endian.h

Currently failing to build for me with this error:
.\psycopg3\types\endian.h(71): fatal error C1083: Cannot open include file: 'sys/param.h': No such file or directory

asqui committed:
test_ssl_in_use(): Adapt for Windows

dvarrazzo committed:
Setting autocommit made thread safe

Using a setter on sync connections, a function on async connections.
Should review why this approach was abandoned for the client encoding,
because it seems working alright.

dvarrazzo committed:
Connection.encoding made writable again

I'm so flipflopping on this...

asqui committed:
First-cut implementation of connection.transaction()

asqui committed:
connection: Move lock acquisition out of _exec_commit_rollback()

dvarrazzo committed:
Connection encoding renamed to client_encoding

More precise and consistent with set_client_encoding()

dvarrazzo committed:
Create two different python distributions in the project

The psycopg3 package is pure Python and doesn't depend on Cython, the
psycopg3_c package is entirely optional and depends on the former.

Tox had to be split, because it gets confused by the presence of the
top-level pyproject.toml, which is there only to configure black.

asqui committed:
Tests: Create svcconn with autocommit on

So that you don't deadlock the test run if you forget to call commit()

e.g. If you use svcconn to create a test table for use by tests, and
forget to call svcconn.commit(), the first test to (try to) touch the
table will block forever waiting on the svcconn transaction to commit.

asqui committed:
mypy passing

asqui committed:
Implement review feedback from asqui#1

asqui committed:
Make conn._savepoints Optional[List[str]]

Implementing review feedback from asqui#1

asqui committed:
Add create_test_table fixture factory; extract transaction tests

This mostly reverts test_connection.py back to starting state

asqui committed:
Remove create_test_table fixture factory

asqui committed:
Use assert_[not_]in_transaction() helper methods

asqui committed:
Move helper functions to the top of test module

asqui committed:
Add connection.transaction(savepoint_name) argument

asqui committed:
Add connection.transaction(force_rollback) argument

To unconditionally discard all changes at end of context.

asqui committed:
Add ability to rollback transactions explicitly

...in the middle of a block, by raising a Rollback exception.

asqui committed:
fixup! Remove create_test_table fixture factory

_Really_ remove create_test_table fixture factory

dvarrazzo commented on psycopg3/connection.py:
```diff
@@ -306,8 +306,10 @@ def _exec_command(self, command: bytes) -> None:
-    def transaction(self) -> Transaction:
-        return Transaction(self)
+    def transaction(
+        self, savepoint_name: str = None, force_rollback: bool = False
```
I don't think this passes Mypy check, right? It should be `Optional[str]`

asqui replied:
Oops. Forgot about mypy at the end -- I've pushed a commit for mypy compliance now.

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -4,7 +4,7 @@
 
 # Copyright (C) 2020 The Psycopg Team
 
-# TODO: Should this module be _underscored since it's implementation detail?
```
I guess underscoring imports was something people used to do when they thought `import *` was a good idea :)

dvarrazzo replied:
(and even in that case, before `__all__` was commonplace)

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
+    enclosing transactions contexts up to and including the one specified.
+    """
+
+    def __init__(self, transaction: "Transaction" = None) -> None:
```
`transaction` is optional, and I think `-> None` is not required for c'tors

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
+    def __init__(self, transaction: "Transaction" = None) -> None:
+        self.transaction = transaction
+
+    def __str__(self) -> str:
```
I don't expect people to call `__str__` (i.e. to routinely print this error in some message of theirs) but `__repr__` is surely useful. I suggest it to make it a representation of the constructor, so `f"{self.__class__.__name__}({self.transaction:r})"` could be good

asqui replied:
Agree. And I think the same goes for `Transaction.__str__()` too?

dvarrazzo replied:
Yup, I think I dropped a note there too.

asqui replied:
Would you say `__repr__()` warrants tests?

Since I can see you are quite strict about what is part of the interface (and therefore should have test coverage) and what isn't.

Also, I think [\_\_qualname\_\_](https://docs.python.org/3/library/stdtypes.html#definition.__qualname__) is the new canonical replacement for `__name__`.

dvarrazzo replied:
I would be ok without tests for repr. Let's just assume it won't explode.

> Also, I think `__qualname__` is the new canonical replacement for `__name__`.

If it looks the best, sure

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
-        self._outer_transaction = False
         self._savepoint_name: Optional[bytes] = None
+        if savepoint_name is not None:
+            self._savepoint_name = savepoint_name.encode("ascii")
```
Just in case we end up allowing user access to the savepoint names, this might be `self.connection.codec.encode(savepont_name)[0]`

dvarrazzo replied:
...ah it seems we do expose it. In this case yes, please use the connection encoding to encode this string

asqui replied:
Ah good, I did feel slightly bad about the hard-coded ascii encodings but hadn't got to thinking about the "proper" way to do this. Thanks!

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
-
-                self._conn._exec_command(b"begin")
+                self._log_and_exec_command(b"BEGIN")
             else:
```
I would make it `elif transaction_stauts == INTRANS` and fail with an unexpected state on `else`

asqui replied:
I've made a note of this one on the todo list too, and will revisit once this PR is closed. (I had some wip tests that I think relate to this and raise some design questions, so I want to have those discussions after I close down this PR.)

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
-                self._savepoint_name = None
-
-                self._conn._exec_command(b"begin")
+                self._log_and_exec_command(b"BEGIN")
```
I think we should be consistent with uppercase and lowercase of the statements we generate. In other places of the adapter I have used lowercase. I don't like to SCREAM at my databases, they may get upset

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
-                self._savepoint_name = None
-
-                self._conn._exec_command(b"begin")
+                self._log_and_exec_command(b"BEGIN")
```
The fact this method logs is not such an important matter to deserve it mentioned in the method name. If we think logging can be useful we might log in `_exec_command`

asqui replied:
Agreed. Would be good to log all commands that the connection executes "on behalf of" the client code.

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
-                self._savepoint_name = b"tx_savepoint_%i" % (
-                    len(self._conn._savepoints) + 1
+                if self._savepoint_name is None:
+                    self._savepoint_name = b"tx_savepoint_%i" % (
```
Bikeshed: Why so verbose? what about `b"s%02d"`? (s01, s02 etc)

asqui replied:
Readability, mainly, but I don't feel strongly. I guess the only time you see this value is when you look at `tx.savepoint_name` or in log messages that say `[RELEASE] SAVEPOINT ...` both of which make it clear that it's a savepoint name, so I guess  the prefix is not adding much readability.

I based this on sqlalchemy, which names its savepoints `sa_savepoint_n`.

If we want to be more succinct, do we even need the zero-padding?

dvarrazzo replied:
If that's the colour of your bike shed, sure :) We don't need them sortable.

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -50,29 +88,48 @@ def __exit__(
                     self._conn._savepoints = None
             else:
+                # Rollback changes made in the transaction context
+                if exc_type is Rollback:
```
`issubclass(exc_type, Rollback)` would be better here, which is similar to what `except Rollback` would do.

asqui replied:
I'll use `if isinstance(exc_val, Rollback):` to match what I'm doing below to satisfy mypy (since it doesn't know the relationship between `exc_type` and `exc_val`).

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -50,29 +88,48 @@ def __exit__(
+                    self._log_and_exec_command(
+                        b"ROLLBACK TO SAVEPOINT " + self._savepoint_name
                     )
                 if self._outer_transaction:
```
Isn't this an `elif`?

asqui replied:
Both blocks should execute in what I'm calling "use case 2", when there is no outer transaction in progress yet, _and_ the user has specified a particular `savepoint_name` -- here the context will "2. Begin an outer transaction and create a savepoint (if one is named)"

We could optimise slightly because a `ROLLBACK TO SAVEPOINT` (or `RELEASE SAVEPOINT`) immediately followed by a `ROLLBACK` (or `COMMIT`) is redundant, but it will complicate the code (we'd still need to do the correct bookkeeping with the savepoint stack), and we've already said we don't care about the performance hit of issuing both `BEGIN` and `SAVEPOINT` commands (only for use case 2).

asqui replied:
Also, your question highlighted that there was no test coverage for this, which I've now added. Thanks :-)

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -50,29 +88,48 @@ def __exit__(
+
+                if exc_type is Rollback:
+                    if exc_val.transaction in (self, None):
+                        return True  # Swallow the exception
```
Mypy won't be happy of returning true, It will want you to return None

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -50,29 +88,48 @@ def __exit__(
+        self._conn._exec_command(command)
+
+    def __str__(self) -> str:
+        return f"Transaction {self._savepoint_name} on {self._conn}"
```
This is a `__repr__`, not a `__str__`

dvarrazzo commented on tests/test_transaction.py:
```diff
@@ -246,39 +286,172 @@ def test_nested_inner_scope_exception_handled_in_outer_scope(conn, svcconn):
+    3. Create a savepoint (if a transaction is already in progress)
+       either using the name provided, or auto-generating a savepoint name.
+    """
+    with caplog.at_level(logging.DEBUG, logger=transaction._log.name):
```
Logging is not part of the adapter interface, and I wouldn't like tests to break because we change it. If you want to test that the adapter keeps on behave well you can monkeypatch the `conn._exec_command` method

asqui replied:
Nice. I felt a bit icky relying on logging. This is a much better approach. Thanks.

dvarrazzo replied:
pytest's monkeypatching should be enough for the task: https://docs.pytest.org/en/stable/monkeypatch.html

asqui replied:
I used `mock.patch` through force-of-habit:
https://github.com/asqui/psycopg3/pull/2/commits/8f758bb63f2615a62c868e567fa8126c9f6654fb

Any objections, given that `mock` is now in the standard library?

dvarrazzo replied:
Some of the unittest maintainers [have ideas I don't appreciate](https://github.com/psycopg/psycopg2/blob/cecff195fc17a83d593dd62c239aa188883a844e/tests/testutils.py#L63-L72) and I would rather stay clear of it.

I have already used pytest's monkeypatch in other places: unless it's tricky I would prefer to avoid using unittest.

dvarrazzo commented on tests/test_transaction.py:
```diff
@@ -246,39 +286,172 @@ def test_nested_inner_scope_exception_handled_in_outer_scope(conn, svcconn):
+        assert_rows(conn, set())
+
+
+def test_explicit_rollback_of_enclosing_transaction_outer_transaction_unaffected(
```
Agh, so long names that the arguments get pushed the line below? Too long imo: a docstring if you think a shorter name wouldn't explain enough would be better.

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -13,35 +13,73 @@
+        return self._savepoint_name.decode("ascii")
+
+    @property
+    def rollback_exception(self) -> Rollback:
```
Looking at tests and use cases, in the test suite you had to import `Rollback` in order to use it, or the user can use this helper method. I think we can do without it and expose the exception in the transaction class, so that the user can do `raise my_tx.Rollback`, `raise my_tx.Rollback(my_tx)`, `raise my_tx.Rollback(other_tx)`.

asqui replied:
I've not come across this pattern of exposing "relevant types" on another type, with the intent for them to be accessed via instances. What's the problem with importing more types as needed?

I worry that this will introduce extra confusion, particularly in statements like `raise my_tx.Rollback(other_tx)` where two transactions are mentioned but only one is actually relevant.

Also, I don't think it's nice if you have to alias the transaction just for the sake of getting access to the `Rollback` type.

e.g. I prefer this...
```
from psycopg3.transactions import Rollback

with conn.transaction():
    raise Rollback()
```

...rather than this:
```
with conn.transaction() as tx:
    raise tx.Rollback()
```

...and this I think becomes catastrophically confusing:
```
with conn.transaction() as tx1:
    with conn.transaction() as tx2:
        raise tx1.Rollback()
```

I'd propose removing this `rollback_exeption` property -- I don't think it's adding much value and just gives you two ways to do the same thing. `Rollback(tx)` isn't exactly a hardship (however we settle on accessing the `Rollback` type.)

dvarrazzo replied:
If it's confusing that's ok. My problem is with exposing the `psycopg3.transaction` module only to expose `Rollback`. I would rather import it in the top `psycopg3` package. And yes, we can probably do without `rollback_exeption`.

asqui replied:
Agree that we don't want people importing `psycopg3.transaction` (hence why I initially wanted to _underscore it :-))
I think we should expose `Rollback` in `psycopg3.__all__` -- do you agree?

dvarrazzo replied:
I think it's a pretty important object: I wouldn't mind it in the top-level package.

asqui committed:
mypy compliance

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -86,14 +89,15 @@ def __exit__(
         exc_val: Optional[BaseException],
         exc_tb: Optional[TracebackType],
-    ) -> None:
+    ) -> bool:
```
Why does this return bool?

asqui replied:
`__exit__` returns a boolean to signal whether the exception should be propagated or not, per the [Python Data Model](https://docs.python.org/3/reference/datamodel.html#object.__exit__) -- so that we can swallow the `Rollback()` exception, where appropriate.

Did I misunderstand the question?

dvarrazzo replied:
No it's me who didn't know the behaviour of returning True from `__exit__` :disappointed: This is all ok (as long as mypy is ok with all the possible exit paths)

dvarrazzo commented on psycopg3/transaction.py:
```diff
@@ -123,13 +128,14 @@ def __exit__(
 
     def __str__(self) -> str:
-        return f"Transaction {self._savepoint_name} on {self._conn}"
+        return f"Transaction {self._savepoint_name!r} on" f" {self._conn}"
```
Split string?

asqui committed:
Refactor test_named_savepoints...

...isolate the log checking to a helper

asqui committed:
Refactor test_named_savepoints...

...assert on _exec_command() calls rather than log messages

asqui committed:
Refactor test_named_savepoints...

...assert on __exit__ too

asqui committed:
Add 2nd test_named_savepoints to cover exceptions

asqui committed:
Implement general code review feedback

asqui committed:
Factor out assert_commands_issued()

...to reduce duplication.

asqui committed:
Add test coverage for asserts

dvarrazzo committed:
Update pytest to v6

dvarrazzo committed:
Dropped use of plain functions to define dumpers

dvarrazzo committed:
Added oid property on Dumper object

dvarrazzo committed:
Added Transformer.types_sequence

Not entirely sure yet if it will be needed, but this allows to free
dump_sequence() from returning the oids as second argument.

Dropped Transformer.get_dump_function() too.

dvarrazzo committed:
Dropped oids from Transform.dump_sequence() return value

dvarrazzo committed:
Dropped MaybeOid type and optional second return value from Dumper.dump()

dvarrazzo committed:
Dropped Transformer.dump()

dvarrazzo committed:
Dropped Transform.dump_sequence(), Transform.types_sequence()

dvarrazzo committed:
All loader functions converted to classes

dvarrazzo committed:
Dropped support for loader functions

asqui committed:
Raise ValueError if savepoint_name is empty

dvarrazzo committed:
Creating a Cython Loader with the same interface as Python

Added C implementation of the BinaryByteaLoader to test the interface

asqui committed:
xfail reminder to escape savepoint names

asqui committed:
Allow multiple commands in _exec_command()

asqui committed:
Release savepoints after rolling back to them

asqui committed:
Misc test refinements

asqui committed:
Don't be so eager to execute 'begin'

Only execute an automatic 'begin' statement if the TransactionStatus is
IDLE, so that a failure to execute 'begin' doesn't overshadow a failure
to execute a user query.

dvarrazzo committed:
Allow overriding of Py loaders by C loaders

Only registering our toy bytea loader for the moment, other will follow
a similar pattern.

dvarrazzo committed:
Added guard to avoid importing psycog3_c without psycopg3

Doing so result in mysterious import error.

dvarrazzo committed:
Dropped unused Transform._oids

dvarrazzo committed:
Use a C fast path to use the Loader.cload where available

dvarrazzo committed:
All C loaders converted to the PyxLoader class

dvarrazzo committed:
PyxLoader renamed to CLoader

dvarrazzo committed:
Dropped Transformer.load()

dvarrazzo committed:
Dropped Transformer.get_load_function

dvarrazzo committed:
Transformer.lookup_loader/dumper dropped from interface

dvarrazzo committed:
Transformer._lookup_dumper/loader dropped altogether

dvarrazzo committed:
Ignore html files generated by Cython

dvarrazzo committed:
Added Escaping.escape_literal

dvarrazzo committed:
Added Escaping.escape_identifier

dvarrazzo committed:
Added Escaping.escape_string

dvarrazzo committed:
More complete string escaping test

dvarrazzo commented on tests/test_transaction.py:
```diff
@@ -322,6 +323,14 @@ def test_named_savepoint_empty_string_invalid(conn):
         conn.transaction(savepoint_name="")
 
 
+@pytest.mark.xfail(raises=OperationalError, reason="TODO: Escape sp names")
```
I've added [`pq.Escaping.escape_literal()`](https://github.com/psycopg/psycopg3/commit/a379f256687029c1d475adaaec5e3683dd8e2a20) in master: you can use it to sanitise the savepoint name.

asqui replied:
I'll do this after I rebase the branch from master, but I want to close out this PR first as it's quite long and confusing, so I want to draw a line under it now without introducing more complexity with rebase/merge commits.

I've made a note of this on the todo list at  https://github.com/asqui/psycopg3/projects/1 so that this doesn't get forgotten.

dvarrazzo committed:
Test fixed to not assume PID < 65535

dvarrazzo committed:
Test and dev dependencies expressed as extras

dvarrazzo committed:
Flake8 configured to deal with the psycopg3 dir

dvarrazzo committed:
Merge branch 'simplify-adapt'

dvarrazzo committed:
Fixed dev mode readme instructions

dvarrazzo committed:
Trailing comma dropped from where black would reformat

The changes are detected by black 20.8b1

asqui commented:
Ok, slightly dropped the ball on this stream of work for the past... nearly 3 months :-o

Coming back to it now and will pick up where we left off.

asqui committed:
fixup! Release savepoints after rolling back to them

Update some tests that were missed with this change.

asqui committed:
Implement more code review feedback from PR asqui#2

* Remove rollback_exception property
* Expose Rollback exception type directly in psycopg3.__all__
* Replace user-reachable `assert` cases with `raise ProgrammingError`
* Use isinstance() rather than `exc_type is Rollback`
* Remove use of `unittest.mock`

asqui commented:
I've pushed 3 commits that I believe address all open questions on this PR.

NB: GitHub doesn't seem to differentiate between when a commit was _authored_ and when it was _pushed_ and incorrectly added commit e561704 (Don't be so eager to execute 'begin') -- which I authored on 7 Aug, but only just pushed -- to an earlier list of commits in this PR. I don't believe this is a change you will have seen previously, so please take a look at that commit too.

I think next steps are to:
1. Merge master into the asqui/transactions branch to incorporate latest changes since this branch was originally created in July!
2. Create a PR to cherry-pick any general changes that don't directly relate to transaction implementation and merge these straight to master. (Such as the aforementioned e561704 (Don't be so eager to execute 'begin'), which is a change to `Connection._start_query`)
3. Proceed with follow-ups listed at https://github.com/asqui/psycopg3/projects/1
4. Create a third and final PR to discuss final design aspects (which I hope will be much shorter-lived and less detailed than the previous two)
5. Once PR asqui#3 is settled, squash everything on the branch and create a PR to bring this into psycopg3/master as a single "Implement transactions" commit.

Let me know your thoughts.

dvarrazzo commented:
> Ok, slightly dropped the ball on this stream of work for the past... nearly 3 months :-o

No worries, I haven't been a juggernaut of code either in the last months. Thankfully I'm unemployed again now and can start putting effort in the project again :D

dvarrazzo commented:
> Create a PR to cherry-pick any general changes that don't directly relate to transaction implementation

I can try and do this on master, as a start, and rebase your branch.

dvarrazzo committed:
Fixed hacking instructions in readme

dvarrazzo committed:
Transaction module moved to the new package position

asqui commented:
> > Create a PR to cherry-pick any general changes that don't directly relate to transaction implementation
> 
> I can try and do this on master, as a start, and rebase your branch.

Sounds good. Let me know when you're done. I'll hold-off on doing any further work until the branch is rebased to avoid any confusion and further changes that would need to be rebased.

dvarrazzo committed:
Make mypy happy

dvarrazzo commented:
I have rebased on master the common things and push-forced here what should be only tx-related.

The last 2 commits are the hacks to make the tests run on windows, and break Linux. The state before them seem to compile and test ok.

asqui commented:
I've squashed the commits relating to transactions and pushed to a new temporary branch `transactions-pr2-end` to form the base for the next (and final!?) PR.

Commits relating to windows compatibility are on the `windows-compatibility` branch and require some thought but I'll continue procrastinating on that :-)

dvarrazzo commented:
Thank you very much, @asqui!

dvarrazzo commented:
I have opened [a nested-transactions branch](https://github.com/psycopg/psycopg3/tree/nested-transactions) in the main repos. I've rebased it on master, addressed a few changed that happened there in the meantime (connection.codec is gone, now connection.client_encoding is a Python codec name), and because I had added the `psycopg3.sql` module in the meantime, now it is easy to escape the savepoint names.

I will leave the branch open, because support for async transactions is missing.

Thank you very much for your work!
